apps/designer/app/canvas/shared/use-drag-drop.ts

```diff
@@ -0,0 +1,273 @@
+      );
+
+      if (!data) {
+        return getDefaultDropTarget();
+      }
+
+      if (data.id === dropTarget.data.id) {
+        return dropTarget;
+      }
+
+      const element = data && document.getElementById(data.id);
+
+      if (!element) {
+        return getDefaultDropTarget();
+      }
+
+      return { data, element };
+    },
+
+    onDropTargetChange(dropTarget) {
+      state.current.dropTarget = dropTarget;
+      publish<"dropTargetChange", DropTargetChangePayload>({
+        type: "dropTargetChange",
+        payload: {
+          rect: dropTarget.rect,
+          placement: dropTarget.placement,
+          position: dropTarget.indexWithinChildren,
+          instanceId: dropTarget.data.id,
+          instanceComponent: dropTarget.data.component,
+        },
+      });
+    },
+  });
+
+  const dragHandlers = useDrag<Instance>({
+    isDragItem(element) {
+      if (rootInstance === undefined || element.id === "") {
+        return false;
+      }
+
+      const instance = findInstanceById(rootInstance, element.id);
+
+      if (instance === undefined) {
+        return false;
+      }
+
+      // We can't drag if we are editing text
+      if (instance.id === textEditingInstanceId) {
+        return false;
+      }
+
+      // Cannot drag root
+      if (instance.id === rootInstance.id) {
+        return false;
+      }
+
+      return instance;
+    },
+    onStart({ data: instance }) {
+      state.current.dragItem = instance;
+
+      autoScrollHandlers.setEnabled(true);
+
+      publish<"dragStart", DragStartPayload>({
+        type: "dragStart",
+        payload: {
+          origin: "canvas",
+          dragItem: instance,
+        },
+      });
+    },
+    onMove: (poiterCoordinate) => {
+      dropHandlers.handleMove(poiterCoordinate);
+      autoScrollHandlers.handleMove(poiterCoordinate);
+    },
+    onEnd() {
+      dropHandlers.handleEnd();
+      autoScrollHandlers.setEnabled(false);
+
+      publish<"dragEnd", DragEndPayload>({
+        type: "dragEnd",
+        payload: { origin: "canvas" },
+      });
+
+      const { dropTarget, dragItem } = state.current;
+
+      if (dropTarget && dragItem) {
+        publish<
+          "reparentInstance",
+          {
+            instanceId: Instance["id"];
+            dropTarget: { instanceId: Instance["id"]; position: number };
+          }
+        >({
+          type: "reparentInstance",
+          payload: {
+            instanceId: dragItem.id,
+            dropTarget: {
+              instanceId: dropTarget.data.id,
+              position: dropTarget.indexWithinChildren,
+            },
+          },
+        });
+      }
+
+      state.current = { ...initialState };
+    },
+  });
+
+  // We have to use useLayoutEffect to setup the refs
+  // because we want to use <body> as a root.
+  // We prefer useLayoutEffect over useEffect
+  // because it's closer in the life cycle to when React noramlly calls the "ref" callbacks.
```

Comment:
this literally doesn't matter, the point of using useLayoutEffect over useEffect is to be called after the dom node is rendered but synchronously, so that you can read layout from dom without causing a reflow, in this case it literally doesn't matter afaik (I haven't tested), I would just use an effect

Reply:
It does matter in our case.

If I use `useEffect` in the following place `rootRef.current` isn't initialised in time:

https://github.com/webstudio-is/webstudio-designer/blob/c8f5b8a341865e9d357b6ca11af3fbc23e869a72/packages/design-system/src/components/primitives/dnd/use-drag.ts#L138-L146